heat/cluster/spectral.py

```diff
@@ -0,0 +1,309 @@
+import torch
+import math
+import numpy as np
+import heat as ht
+import time
+
+
+def laplacian(S, norm=True, mode="fc", upper=None, lower=None):
+    """
+    Calculate the graph Laplacian from a similarity matrix
+
+    Parameters
+    ----------
+    S : ht.DNDarray
+        quadrdatic, positive semidefinite similarity matrix, encoding similarity metrices s_ij between data samples i and j
+    norm : bool
+        Whether to calculate the normalized graph Laplacian
+        The unnormalized graph Laplacian is defined as L = D - A, where where D is the diagonal degree matrix and A the adjacency matrix
+        The normalized graph Laplacian is defined as L_norm = D^(-1/2) L D^(-1/2) = I - D^(-1/2) A D^(-1/2)
+    mode : "fc", "eNeighbour"
+        How to calculate adjacency from the similarity matrix
+        "fc" is fully-connected, so A = S
+        "eNeighbour" is the epsilon neighbourhood, with A_ji = 1 if S_ij >/< lower/upper else 0; for eNeighbour an upper or lower boundary needs to be set
+    upper : float
+        upper boundary for adjacency calculation, using A_ji = 1 if S_ij < upper else 0
+    lower : float
+        upper boundary for adjacency calculation, using A_ji = 1 if S_ij > lower else 0
+    Returns
+    -------
+    L : ht.DNDarray
+
+    """
+
+    if mode == "eNeighbour":
+        if (upper is not None) and (lower is not None):
```

Comment:
brackets here not necessary I believe. Is this an auto addition by black?

Reply:
no, I do this for better readibility (C++ habits die hard). But I can remove them

Reply:
Would be good thanks